### PR TITLE
HPCC-28443 Resolve roxie hostnames on demand in ESP services

### PR DESCRIPTION
**Pull request by wangkx**
1. Add new CConnectionSettings to store roxie communication settings.
2. Add new roxie control interfaces to resolve roxie hostnames on demand.
3. Revise ESP services to use the communication settings and new control interfaces.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/clients/roxiecontrol.cpp [hunk 1]
```diff
@@ -18,12 +18,81 @@
...
+        FQDNs.append(name);
+    }
+}
+
+void CConnectionSettings::parseIPRange(char *targets, char *dash)
+{
+    *dash = 0;
+    int numberAfterDash = atoi(dash+1);
+    char *lastDot = strrchr(targets, '.');
+    *lastDot = 0;
```
Comment (jakesmith) [inline esp/clients/roxiecontrol.cpp:65]: See question above re. why parsing , but if it is necessary, then not good practice to alter memory of owned string like this.

If it's really needed - I assume this range format is used elsewhere. If so, this parsing code should be commoned up in a single hepler caller by both.

File: esp/clients/roxiecontrol.cpp [hunk 2]
```diff
@@ -18,12 +18,81 @@
...
+        unsigned j = random->next() % i;
+        i--;
+        FQDNs.swap(i, j);
+    }
+}
+
+const char *CConnectionSettings::nextFQDN()
+{
+    const char *FQDN = FQDNs.item(nextFQDNIndex);
+    ++nextFQDNIndex %= FQDNs.ordinality();
```
Comment (jakesmith) [inline esp/clients/roxiecontrol.cpp:92]: related to above - I don't see how this fits in with the schema of a roxie service definition.

**Review by jakesmith — COMMENTED** [review]
@wangkx - needs more discussion I think.

The new class (CConnectionSettings) and the maps (e.g. connSettingMap), and conditional code looks like it is in some ways replicating the pre-existing functionality that the smart socket factory implements, but, for example, there should be no ip ranges (or ips) in a roxie service @name that it's parsing.

If it was a new implementation that supported both bare-metal roxie definitions and k8s service definitions and coped with both, thaat would make sense - but the new class is only used by containerized and there's quite of #ifdef _CONTAERINIZED blocks #else bare-metal - which makes the code more complicated and harder to maintain.

Could both bare-metal and containerized use CConnectionSettings ? 

But also, the primary objective here is to avoid cached staatel ep's correct?

It looks like CSmartSocketFactory is already designed to cope with holding hosts (if that's what it is passed via config) and re-resolving them automatically based on a configured timeout (via dnsInterval param)
Was there a reason using that wasn't pursued for resolving this?

File: esp/clients/roxiecontrol.cpp [hunk 3]
```diff
@@ -18,12 +18,81 @@
...
+{
+    name.set(service.queryProp("@name"));
+    if (isEmptyString(name))
+        throw makeStringException(-1, "CConnectionSettings: name required for service configuration");
+
+    port = service.getPropInt("@port");
+    if (port == 0)
+        throw makeStringException(-1, "CConnectionSettings: invalid port for service configuration");
+
+    tlsService  = service.getPropBool("@tls");
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/clients/roxiecontrol.cpp:38]: Will fix

File: esp/clients/roxiecontrol.cpp [hunk 4]
```diff
@@ -18,12 +18,81 @@
...
 #include "jmisc.hpp"
 #include "securesocket.hpp"
+#include "jsecrets.hpp"
 
 const unsigned roxieQueryRoxieTimeOut = 60000;
 
 #define EMPTY_RESULT_FAILURE 1200
 #define SECURE_CONNECTION_FAILURE 1201
 
+CConnectionSettings::CConnectionSettings(IPropertyTree &service)
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/clients/roxiecontrol.cpp:28]: Will add the const.

File: esp/smc/SMCLib/TpWrapper.hpp [hunk 5]
```diff
@@ -216,7 +214,11 @@ extern TPWRAPPER_API unsigned getWUClusterInfo(CConstWUClusterInfoArray& cluster
 extern TPWRAPPER_API IConstWUClusterInfo* getWUClusterInfoByName(const char* clustName);
 
 extern TPWRAPPER_API IStringIterator *getContainerTargetClusters(const char* processType, const char* processName);
+#ifndef _CONTAINERIZED
 extern TPWRAPPER_API void initContainerRoxieTargets(MapStringToMyClass<ISmartSocketFactory>& connMap);
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/smc/SMCLib/TpWrapper.hpp:218]: This version is not called now. The code should be removed when separating the code TpContainer.cpp.

File: esp/clients/roxiecontrol.cpp [hunk 6]
```diff
@@ -18,12 +18,81 @@
...
+    if (tlsService)
+        tlsConfig.setown(createTlsClientSecretInfo(service.queryProp("@issuer"), !service.getPropBool("@public"), service.getPropBool("@selfSigned"), service.getPropBool("@caCert")));
+
+    readFQDNs();
+}
+
+void CConnectionSettings::readFQDNs()
+{
+    char *targets = (char *) name.get();
+    char *dash = strrchr(targets, '-');
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/clients/roxiecontrol.cpp:48]: The @name is copied from old code which parses an IP range in CSmartSocketFactory. Do we allow an IP range configured for containerized roxie service?

**Comment by wangkx** [comment]
I do agree we need more discussion. 
To my understand, the primary objective here is to avoid cached ep's. I haven't found where the CSmartSocketFactory stores original host settings. So, not sure how to reset the ep's (old IP address) using original host settings in CSmartSocketFactory.
For this PR, the new class (CConnectionSettings) may only be used for containerized roxie. I am still debating whether it should be used to bare metal roxie or not (may be later?). To be used in bare metal roxie, the class has to be upgrade to support many bare metal features (ex. support '|' separated string for multiple host/port settings, etc.).

**Comment by wangkx** [comment]
@jakesmith looks like that the name is stored inside the SmartSocketEndpoint. If the dnsInterval is set to 0, no cached IP will be used by the CSmartSocketFactory. So, the cached IP issue may be fixed by changing the dnsInterval to 0 for K8s environment. Any comment?

**Comment by wangkx** [comment]
I am testing the 0 dnsInterval. It does not help for the calls from the WsEcl and WsSMC. Will do more research.

**Comment by wangkx** [comment]
The ESP services use the roxie connections through the nextEndpoint() call. The nextEndpoint() does not use the dnsInterval. In my latest commit, a new nextEndpointNoCache() is created and used. @afishbeck @jakesmith please check.

File: system/jlib/jsmartsock.cpp [hunk 7]
```diff
@@ -329,6 +329,17 @@ SocketEndpoint& CSmartSocketFactory::nextEndpoint()
     return (ss->ep);
 }
 
+SocketEndpoint& CSmartSocketFactory::nextEndpointNoCache()
```
Comment (afishbeck) [inline system/jlib/jsmartsock.cpp:332]: Perhaps you should just make nextEndpoint() respect the dnsinterval by adding?:

```
ss->checkHost(dnsInterval);
return (ss->ep);
```

Comment (afishbeck) [reply system/jlib/jsmartsock.cpp:332]: And also go back to defaulting the dnsinterval on containerized to 0.

**Comment by wangkx** [comment]
@afishbeck I have made the changes related to the nextEndpoint().

File: system/jlib/jsmartsock.cpp [hunk 8]
```diff
@@ -326,6 +326,9 @@ SocketEndpoint& CSmartSocketFactory::nextEndpoint()
     SmartSocketEndpoint *ss=nextSmartEndpoint();
     if (!ss)
         throw createSmartSocketException(0, "smartsocket failed to get nextEndpoint");
+
+    synchronized block(lock);
+    ss->checkHost(dnsInterval);
```
Comment (jakesmith) [inline system/jlib/jsmartsock.cpp:331]: I think should move this inside nextSmartEndpoint, but make it conditional,
e.g. nextSmartEndpoint(bool validate)

call it with true here, and look at other places that call it, which manually call checkHost themselves, so wouldn't need to if nextSmartEndpoint has done already.

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks a lot simpler/smaller change.

File: system/jlib/jsmartsock.cpp (hunk 8 above)
Comment (wangkx) [reply system/jlib/jsmartsock.cpp]: Moved

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 9]
```diff
@@ -227,7 +227,7 @@ void initContainerRoxieTargets(MapStringToMyClass<ISmartSocketFactory> &connMap)
         if (isEmptyString(target) || isEmptyString(service.queryProp("@name"))) //bad config?
             continue;
 
-        Owned<ISmartSocketFactory> sf = new WsEclSocketFactory(service, false, true, nullptr, (unsigned) -1);
+        Owned<ISmartSocketFactory> sf = new WsEclSocketFactory(service, false, true, nullptr, 0);
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/ws_ecl/ws_ecl_service.cpp:230]: The underlying OS should be doing some level of caching.  Not specific to these particular services, but especially if we do cache though we need to make sure we have retries in which the retry does try to re-resolve the endpoint before new connection attempt. 

I'm not really sure what all the reasons are that IPs will change, but we probably don't want windows where we will fail when they do.

**Comment by wangkx** [comment]
@jakesmith could you please review my changes?

File: esp/services/ws_ecl/ws_ecl_service.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply esp/services/ws_ecl/ws_ecl_service.cpp:230]: >I'm not really sure what all the reasons are that IPs will change,

I think the main one this issue was related to, was in K8s, where a load-balanced service name resolved to a new IP.

[the remark this answers is not in the document]
Comment (jakesmith) [reply esp/services/ws_ecl/ws_ecl_service.cpp:230]: >but we probably don't want windows where we will fail when they do.

I'm not really sure how the smart socket layer is suppsoed to work,
but it looks like it runs a background thread that together with retry/retryInterval, periodically checks the sockets via checkHost(dnsInterval) and actually tried to connect to it (and marks a status).
I'm not sure what checks that status though.

**Review by jakesmith — APPROVED** [review]
@wangkx - looks good afaics.

File: esp/services/ws_ecl/ws_ecl_service.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/ws_ecl/ws_ecl_service.cpp:230]: Part of the problem with smart socket fully managing addresses is that it is used in two different ways.. 1. smart socket is directly asked to connect.  2. smart socket is asked to provide an address that another library or function will connect to.

In these cases http routines are used to connect after building a url around the address.

**Review by afishbeck — APPROVED** [review]

**Comment by wangkx** [comment]
@ghalliday this PR is ready to be merged.

**Pull request merged**